client-v2/src/shared/selectors/__tests__/shared.spec.js

```diff
@@ -0,0 +1,141 @@
+import {
+  createArticleFromArticleFragmentSelector,
+  externalArticleFromArticleFragmentSelector,
+  createArticlesInCollectionGroupSelector
+} from '../shared';
+
+const state = {
+  collections: {
+    c1: {
+      groups: ['group1', 'group2'],
+      articles: {
+        live: ['af1', 'af2']
+      }
+    },
+    c2: {
+      groups: ['group1'],
+      articles: {
+        draft: ['af3', 'af4']
+      }
+    }
+  },
+  externalArticles: {
+    ea1: {
+      id: 'ea1',
+      headline: 'Example external article'
+    }
+  },
+  articleFragments: {
+    af1: {
+      uuid: 'af1',
+      id: 'ea1',
+      frontPublicationDate: 1,
+      publishedBy: 'A. N. Author',
+      meta: {
+        group: '0'
+      }
+    },
+    afWithInvalidReference: {
+      uuid: 'afWithInvalidReference',
+      id: 'invalid',
+      meta: {}
+    },
+    af2: {
+      uuid: 'af2',
+      meta: {
+        group: '1'
+      }
+    },
+    af3: {
+      uuid: 'af3',
+      meta: {
+        group: '0'
+      }
+    },
+    af4: {
+      uuid: 'af4',
+      meta: {
+        group: '0'
+      }
+    }
+  }
+};
+
+describe('Shared selectors', () => {
+  describe('createExternalArticleFromArticleFragmentSelector', () => {
+    it('should create a selector that returns an external article referenced by the given article', () => {
+      expect(externalArticleFromArticleFragmentSelector(state, 'af1')).toEqual(
+        state.externalArticles.ea1
+      );
+      expect(
+        externalArticleFromArticleFragmentSelector(state, 'invalid')
+      ).toEqual(null);
+    });
+  });
+
+  describe('createArticleFromArticleFragmentSelector', () => {
+    it('should derive an Article from an ArticleFragment and an ExternalArticle, using an ArticleFragment id', () => {
+      const selector = createArticleFromArticleFragmentSelector();
+      expect(selector(state, 'af1')).toEqual({
+        id: 'ea1',
+        headline: 'Example external article',
+        group: '0'
+      });
+    });
+    it('should handle a case where the referenced ArticleFragment does not exist', () => {
+      const selector = createArticleFromArticleFragmentSelector();
+      expect(selector(state, 'invalid')).toEqual(null);
+    });
+    it('should handle a case where the ExternalArticle referenced by the ArticleFragment does not exist', () => {
+      const selector = createArticleFromArticleFragmentSelector();
+      expect(selector(state, 'afWithInvalidReference')).toEqual(null);
+    });
+  });
+
+  describe('createArticlesInCollectionGroupSelector', () => {
```

Comment:
We are missing a test for when the meta block is entirely missing from the article.

Reply:
Excellent point! Also a test for the idea of articles without a 'meta' key belonging to group 0. I'll add both.